### PR TITLE
Add game Bulls and Cows

### PR DESCRIPTION
**Pull request by EkaterinaFarygina**
Game generates secret code with desired length (from 1 to 36) and desired number of possible symbols (from 1 to 36).

You should guess code. If symbol is presented in secret code and placed in the same position as in the secret code, then it is a bull. If a symbol is presented in secret code but placed not in the same position, then it is a cow.

You can win the game only if all symbols are bulls.

### CODE REVIEW

File: src/bullscows/BullsAndCows.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        String cowFormat = cows == 1 ? "cow" : "cows";
+        String bullFormat = bulls == 1 ? "bull" : "bulls";
+        if (bulls == 0 || cows == 0) {
+            if (bulls == 0 && cows == 0) {
+                System.out.print("Grade: None\n");
+            } else if (bulls == 0) {
+                System.out.printf("Grade: %d %s\n", cows, cowFormat);
+            } else {
+                System.out.printf("Grade: %d %s\n", bulls, bullFormat);
+            }
```
Comment (pavel-krizskii) [inline src/bullscows/BullsAndCows.java]: Can be simplified without outer if:
```
            if (bulls == 0 && cows == 0) {
                System.out.print("Grade: None\n");
            } else if (bulls == 0) {
                System.out.printf("Grade: %d %s\n", cows, cowFormat);
            } else if (cows == 0) {
                System.out.printf("Grade: %d %s\n", bulls, bullFormat);
            } else {
                System.out.printf("Grade: %d %s and %d %s\n", bulls, bullFormat, cows, cowFormat);
            }
```

File: src/bullscows/BullsAndCows.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        } else {
+            System.out.printf("Grade: %d %s and %d %s\n", bulls, bullFormat, cows, cowFormat);
+        }
+    }
+
+    boolean isCorrect(int length) {
+        return bulls == length;
+    }
```
Comment (pavel-krizskii) [inline src/bullscows/BullsAndCows.java]: minor: can be replaced with `getBulls() == length` in places where we use it

File: src/bullscows/BullsAndCowsCalculator.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+public class BullsAndCowsCalculator {
+    private final String secretCode;
+
+    BullsAndCowsCalculator(String secretCode) {
+        this.secretCode = secretCode;
+    }
+
+    void displaySecretCode() {
+        System.out.println("The secret code is " + secretCode + ".");
+    }
```
Comment (pavel-krizskii) [inline src/bullscows/BullsAndCowsCalculator.java]: remove me

File: src/bullscows/BullsAndCows.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    private final int cows;
+
+    BullsAndCows(int bulls, int cows) {
+        this.bulls = bulls;
+        this.cows = cows;
+    }
+
+    void displayResult() {
+        String cowFormat = cows == 1 ? "cow" : "cows";
+        String bullFormat = bulls == 1 ? "bull" : "bulls";
```
Comment (pavel-krizskii) [inline src/bullscows/BullsAndCows.java]: minor: final

File: src/bullscows/BullsAndCowsCalculator.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    BullsAndCows calculateBullsAndCows(String inputCode) {
+        var bulls = 0;
+        var cows = 0;
+        for (int i = 0; i < inputCode.length(); i++) {
+            var currentNumber = inputCode.substring(i, i+1);
```
Comment (pavel-krizskii) [inline src/bullscows/BullsAndCowsCalculator.java]:
```
            final var currentDigit = inputCode.substring(i, i + 1);
            final var positionInCode = secretCode.indexOf(currentDigit);
            if (positionInCode == i) {
                bulls++;
            } else if (positionInCode != -1) {
                cows++;
            }
```

File: src/bullscows/SecretCodeGenerator.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
+package bullscows;
+
+import java.util.Arrays;
+import java.util.Random;
+
+public class SecretCodeGenerator {
+
+    int inputLengthOfCode() {
```
Comment (pavel-krizskii) [inline src/bullscows/SecretCodeGenerator.java]: Can we move all input methods (`inputLengthOfCode`, `inputNumberOfSymbols`) to `InputScanner`?

File: src/bullscows/SecretCodeGenerator.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            for (int i = 10; i < numberOfSymbols; i++) {
+                values[i] = (char) (87 + i);
+            }
+        }
+        return values;
+    }
+
+    private void shuffleArrayOfSymbols(char[] values) {
+        final var random = new Random();
+        char container;
```
Comment (pavel-krizskii) [inline src/bullscows/SecretCodeGenerator.java]: we can combine `char container` definition and initialisation below

File: src/bullscows/SecretCodeGenerator.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        for (int i = 0; i < 300; i++) {
+            int randomIndex1 = random.nextInt(values.length);
+            int randomIndex2 = random.nextInt(values.length);
+            container = values[randomIndex1];
```
Comment (pavel-krizskii) [inline src/bullscows/SecretCodeGenerator.java]: minor: why is it called `container`?

File: src/bullscows/SecretCodeGenerator.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        } else {
+            System.out.printf(" (0-9, a-%c).\n", values[numberOfSymbols - 1]);
+        }
+    }
+
+    String generateSecretCode(int length, int numberOfSymbols) {
+        if (length > 36) {
+            System.out.println("Error: can't generate a secret number with a length of more than 36 " +
+                    "because there aren't enough unique symbols.");
+        }
```
Comment (pavel-krizskii) [inline src/bullscows/SecretCodeGenerator.java]: Do we need to add return here in case of this error?

File: src/bullscows/BullsAndCows.java (hunk 2 above)
Comment (EkaterinaFarygina) [reply src/bullscows/BullsAndCows.java]: Removed. Doesn't used anywhere :)

**Review by pavel-krizskii — APPROVED** [review]
LGTM

**Pull request merged**